controller/src/main/java/org/jboss/as/controller/CapabilityRegistry.java

```diff
@@ -0,0 +1,745 @@
+                    } else {
+                        removed = candidate;
+                    }
+                }
+            }
+
+            if (removed != null) {
+                modified = true;
+            }
+            return removed;
+        } finally {
+            writeLock.unlock();
+        }
+    }
+
+
+    //ImmutableCapabilityRegistry methods
+
+    @Override
+    public boolean hasCapability(String capabilityName, CapabilityContext context) {
+        try {
+            readLock.lock();
+            return findSatisfactoryCapability(capabilityName, context, !forServer) != null;
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public <T> T getCapabilityRuntimeAPI(String capabilityName, CapabilityContext context, Class<T> apiType) {
+        // Here we can't know the dependent name. So this can only be called when resolution is complete.
+        assert resolutionContext.resolutionComplete;
+        try {
+            readLock.lock();
+            RuntimeCapabilityRegistration reg = getCapabilityRegistration(capabilityName, context);
+            Object api = reg.getCapability().getRuntimeAPI();
+            if (api == null) {
+                throw ControllerLogger.MGMT_OP_LOGGER.capabilityDoesNotExposeRuntimeAPI(capabilityName);
+            }
+            return apiType.cast(api);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public Set<CapabilityRegistration> getCapabilities() {
+        try {
+            readLock.lock();
+            return Collections.unmodifiableSet(new HashSet<>(capabilities.values()));
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public Set<CapabilityRegistration> getPossibleCapabilities() {
+        try {
+            readLock.lock();
+            return Collections.unmodifiableSet(new HashSet<>(possibleCapabilities.values()));
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    @Override
+    public ServiceName getCapabilityServiceName(String capabilityName, CapabilityContext context, Class<?> serviceType) {
+        // Here we can't know the dependent name. So this can only be called when resolution is complete.
+        assert resolutionContext.resolutionComplete;
+        try {
+            readLock.lock();
+            RuntimeCapabilityRegistration reg = getCapabilityRegistration(capabilityName, context);
+            RuntimeCapability<?> cap = reg.getCapability();
+            return cap.getCapabilityServiceName(serviceType);
+        } finally {
+            readLock.unlock();
+        }
+    }
+
+    //end ImmutableCapabilityRegistry methods
+
+    /**
+     * Publish the changes to main registry
+     */
+    void publish() {
+        assert publishedFullRegistry != null : "Cannot write directly to main registry";
+
+        if (!modified) {
+            return;
+        }
+        try {
+            writeLock.lock();
```

Comment:
This protects the data of 'this' but a lot of the data that's being modified is 'publishedFullRegistry'.

Since this.modified is being changed, protecting that with this.writeLock is ok, but then publishedFullRegistry.writeLock needs to be taken to protect the publishedFullRegistry data.

I don't see any case where publishedFullRegistry would take its own writeLock and then try to do something that results in trying to take this.writeLock, so there's no risk of deadlock from taking the two locks.